src/libraries/System.Private.CoreLib/src/System/IO/BufferedStream.cs

```diff
@@ -51,6 +52,12 @@ public sealed class BufferedStream : Stream
         private Stream? _stream;                            // Underlying stream.  Close sets _stream to null.
         private byte[]? _buffer;                            // Shared read/write buffer.  Alloc on first use.
         private readonly int _bufferSize;                   // Length of internal buffer (not counting the shadow buffer).
+        /// <summary>
+        /// allows for:
+        /// 1. blocking zero byte reads
+        /// 2. skipping the serialization of async operations that don't use the buffer
+        /// </summary>
+        private readonly bool _actLikeFileStream;
```

Comment:
For (1), can we just make it so that BufferedStream for a read of 0 and nothing buffered issues a read on the underlying stream?  That it doesn't today seems like a bug that we should fix rather than special-casing it.  Obviously any change can break something, but is there anything meaningful that actually breaks if we do that?

For (2), what happens if we don't special-case that and just always do that serialization?

Reply:
> would it be OK If I change this behaviour in a separate PR? Just in case it turns out to be breaking something.

Ok

> we lose quite a lot of perf as we go the slow path (few extra awaits and some other overhead that adds up)

My concern is really that we seem to be creating two largely different implementations here that just happen to share some code: one when used for FileStream, and one when used for everything else.  If the implementation isn't sufficient for FileStream, and we need these complications to bend BufferedStream to FileStream's will, should BufferedFileStreamStrategy just have the buffering logic built into it rather than using BufferedStream?  Then BufferedStream doesn't need all these extra checks and code paths, and the implementation in FileStream can be tailored to just what FileStream needs.